### PR TITLE
add open-llama model with ckpt

### PR DESCRIPTION
**Pull request by s-JoL**
This PR adds a new model called Open-Llama, which is based on Llama's implementation in Transformers. 
In Open-Llama,  emory-efficient attention has been added, resulting in a 30% improvement in training efficiency. Additionally, hidden dropout and attention dropout have been added for better generalization during training.

We have also added two optional features: stable embedding from Bloom and shared input-output vectors from PALM, which have been tested and found to improve training stability and performance.

The following code snippet shows the implementation of memory-efficient attention:

```python
try:
    from xformers import ops as xops
except ImportError:
    xops = None
    print("xformers is not installed correctly.")

if self.config.use_memorry_efficient_attention and xops is not None and self.training:
    attn_weights = None
    query_states = query_states.transpose(1, 2)
    key_states = key_states.transpose(1, 2)
    value_states = value_states.transpose(1, 2)
    attn_output = xops.memory_efficient_attention(
        query_states, key_states, value_states, attn_bias=xops.LowerTriangularMask(), p=self.dropout_prob
    )
```

At the same time, for maximum compatibility, we have made xformers an optional dependency so that the original implementation can still be used for training and inference if it is not installed.

We implemented pre-training of the Llama model based on transformers + accelerate, incorporating the modifications described above.
[Open-Llama](https://github.com/Bayes-Song/Open-Llama/blob/main/README_en.md)

The pre-trained model has already been open-sourced on [s-JoL/Open-Llama-V1](https://huggingface.co/s-JoL/Open-Llama-V1).

ref: https://github.com/huggingface/transformers/pull/22386

cc: @sgugger

### CODE REVIEW

**Comment by sgugger** [comment]
cc @ArthurZucker and @younesbelkada

**Comment by s-JoL** [comment]
Please help me review this pull request. @ArthurZucker  @younesbelkada

**Comment by ArthurZucker** [comment]
Hey! Thanks will review now

File: src/transformers/models/open_llama/modeling_open_llama.py [hunk 1]
```diff
@@ -0,0 +1,922 @@
...
+        self.mlp = OpenLlamaMLP(
+            hidden_size=self.hidden_size,
+            intermediate_size=config.intermediate_size,
+            hidden_act=config.hidden_act,
+            dropout_prob=config.hidden_dropout_prob,
+        )
+        self.input_layernorm = OpenLlamaRMSNorm(config.hidden_size, eps=config.rms_norm_eps)
+        self.post_attention_layernorm = OpenLlamaRMSNorm(config.hidden_size, eps=config.rms_norm_eps)
+
+    def forward(
```
Comment (ArthurZucker) [inline src/transformers/models/open_llama/modeling_open_llama.py:286]: this is missing copeid from statement

File: src/transformers/models/open_llama/modeling_open_llama.py [hunk 2]
```diff
@@ -0,0 +1,922 @@
...
+
+    Parameters:
+        config ([`OpenLlamaConfig`]):
+            Model configuration class with all the parameters of the model. Initializing with a config file does not
+            load the weights associated with the model, only the configuration. Check out the
+            [`~PreTrainedModel.from_pretrained`] method to load the model weights.
+"""
+
+
+@add_start_docstrings(
```
Comment (ArthurZucker) [inline src/transformers/models/open_llama/modeling_open_llama.py:358]: missing copied from statements

File: src/transformers/models/open_llama/modeling_open_llama.py [hunk 3]
```diff
@@ -0,0 +1,922 @@
...
+        # Initialize weights and apply final processing
+        self.post_init()
+
+    def get_input_embeddings(self):
+        return self.embed_tokens
+
+    def set_input_embeddings(self, value):
+        self.embed_tokens = value
+
+    def _prepare_decoder_attention_mask(self, attention_mask, input_shape, inputs_embeds, past_key_values_length):
```
Comment (ArthurZucker) [inline src/transformers/models/open_llama/modeling_open_llama.py:487]: this is also missing copied from

File: tests/models/open_llama/test_modeling_open_llama.py [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+        attention_mask = input_ids.ne(1).to(torch_device)
+        sequence_labels = ids_tensor(
+            [self.model_tester.batch_size, config.num_labels], self.model_tester.type_sequence_label_size
+        ).to(torch.float)
+        model = OpenLlamaForSequenceClassification(config)
+        model.to(torch_device)
+        model.eval()
+        result = model(input_ids, attention_mask=attention_mask, labels=sequence_labels)
+        self.assertEqual(result.logits.shape, (self.model_tester.batch_size, self.model_tester.num_labels))
+
```
Comment (ArthurZucker) [inline tests/models/open_llama/test_modeling_open_llama.py:334]: Do you have any integration tests ? Or this is the code you basically used

**Review by ArthurZucker — COMMENTED** [review]
Thanks for working on this! Seems like the model is overlall very similar, so missing bunch of copied form here and there. Most importantly I dont think we need a new tokenizer, it's still llama tokenizer.

File: src/transformers/models/auto/tokenization_auto.py [hunk 5]
```diff
@@ -244,6 +244,13 @@
                 ),
             ),
             ("oneformer", ("CLIPTokenizer", "CLIPTokenizerFast" if is_tokenizers_available() else None)),
+            (
+                "open-llama",
+                (
+                    "OpenLlamaTokenizer" if is_sentencepiece_available() else None,
+                    "OpenLlamaTokenizerFast" if is_tokenizers_available() else None,
```
[the remark this answers is not in the document]
Comment (s-JoL) [reply src/transformers/models/auto/tokenization_auto.py:251]: indeed, this can be directly reused.

File: src/transformers/models/open_llama/configuration_open_llama.py
[the remark this answers is not in the document]
Comment (s-JoL) [reply src/transformers/models/open_llama/configuration_open_llama.py:1]: I'm concerned that using the default LlamaConfig directly may result in missing parameters and cause errors.

File: src/transformers/models/open_llama/modeling_open_llama.py (hunk 2 above)
Comment (s-JoL) [reply src/transformers/models/open_llama/modeling_open_llama.py:358]: sorry, i did't quite understand how to add the copied from statements for this class, there are slight differences here.

File: src/transformers/models/open_llama/convert_open_llama_weights_to_hf.py
Comment (ArthurZucker) [inline src/transformers/models/open_llama/convert_open_llama_weights_to_hf.py:1]: Same comment here, is this not the same as in the `llama` folder?

File: src/transformers/models/open_llama/modeling_open_llama.py (hunk 2 above)
Comment (ArthurZucker) [reply src/transformers/models/open_llama/modeling_open_llama.py:363]: Ok you can keep it as is!

**Review by ArthurZucker — COMMENTED** [review]
LGTM, waiting for @sgugger's review

File: src/transformers/models/open_llama/convert_open_llama_weights_to_hf.py
Comment (s-JoL) [reply src/transformers/models/open_llama/convert_open_llama_weights_to_hf.py:1]: Thank you for the reminder. This file is identical to the one in Llama, and since I trained directly with Transformers, there is no need for any conversion. I will delete it.

File: docs/source/en/model_doc/open-llama.mdx [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+-->
+
+# Open-Llama
+
+## Overview
+
+The Open-Llama model was proposed in [Open-Llama project](https://github.com/s-JoL/Open-Llama) by community developer s-JoL.
+
+The model is mainly based on LLaMA with some modifications, incorporating memory-efficient attention from Xformers, stable embedding from Bloom, and shared input-output embedding from PLAM.
+And the model is pre-trained on both Chinese and English, which gives it better performance on Chinese language tasks.
```
[the remark this answers is not in the document]
Comment (s-JoL) [reply docs/source/en/model_doc/open-llama.mdx:20]: This is a great suggestion, but currently I have not conducted a complete ablation experiment. I plan to gradually add it to the documentation after conducting the experiment.

File: utils/check_config_docstrings.py [hunk 7]
```diff
@@ -42,6 +42,7 @@
     "VisionEncoderDecoderConfig",
     "VisionTextDualEncoderConfig",
     "LlamaConfig",
+    "OpenLlamaConfig",
```
Comment (sgugger) [inline utils/check_config_docstrings.py:45]: Should be removed as there is a checkpoint for OpenLlama.

File: src/transformers/models/open_llama/configuration_open_llama.py [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+OPEN_LLAMA_PRETRAINED_CONFIG_ARCHIVE_MAP = {
+    "s-JoL/Open-Llama-V1": "https://huggingface.co/s-JoL/Open-Llama-V1/blob/main/config.json",
+}
+
+
+class OpenLlamaConfig(PretrainedConfig):
+    r"""
+    This is the configuration class to store the configuration of a [`OpenLlamaModel`]. It is used to instantiate an
+    Open-Llama model according to the specified arguments, defining the model architecture. Instantiating a
+    configuration with the defaults will yield a similar configuration to that of the Open-Llama-7B.
```
Comment (sgugger) [inline src/transformers/models/open_llama/configuration_open_llama.py:37]: Put the full checkpoint name here and link to the Hub. Example we have for GPT-2: 
```
a similar configuration to that of the [gpt2](https://huggingface.co/gpt2) architecture.
```
It wasn't there for Llama since there is no official checkpoint on the Hub.

**Review by sgugger — APPROVED** [review]
Very clean, thanks a lot for adding this! I have just a comment on the config and default checkpoint.

**Review by sgugger — APPROVED** [review]

**Review by ArthurZucker — APPROVED** [review]

**Comment by sgugger** [comment]
Thanks a lot for your contribution!

**Comment by s-JoL** [comment]
> Thanks a lot for your contribution!
Hello, I have a question, why the open-Llama model cannot be searched in the documentation of transformers? Is there something I forgot to add?

image

**Comment by amyeroberts** [comment]
Hi @s-JoL, thanks for notifying. 

There was an issue in the doc rendering (resolved with [1](https://github.com/huggingface/huggingface-meilisearch/pull/60/files), [2](https://github.com/huggingface/huggingface-meilisearch/pull/61)) leading to some pages not being retrievable in search. Should be working now!

**Comment by PenutChen** [comment]
@s-JoL I noticed that the links pertaining to Open-LLaMA are currently leading to 404 errors. Could you please provide some information on what might have happened?

**Comment by heya5** [comment]
@s-JoL Hi, I can't find a Open-LLaMA checkpoint and I noticed you delete your original repo. What happend? How Can I have a try of Open-LLaMA?

**Comment by PenutChen** [comment]
@heya5 Possibly due to some controversies surrounding this project, the original author has closed the original project.
https://github.com/chenfeng357/open-Chinese-ChatLLaMA/issues/1
